src/Widgets/PopoverWidget.vala

```diff
@@ -340,16 +339,25 @@ public class Network.Widgets.PopoverWidget : Gtk.Grid {
     }
 
     private void update_vpn_connection () {
-        active_vpn_connection = null;
-
+        /* Stupid heuristic for now: at least one connection must be secure to
+         * display the secure badge. */
+        // Reset the current status
+        secure = false;
         nm_client.get_active_connections ().foreach ((ac) => {
-            if (active_vpn_connection == null && ac.vpn) {
-                active_vpn_connection = (NM.VpnConnection) ac;
-
-                secure = active_vpn_connection.get_vpn_state () == NM.VpnConnectionState.ACTIVATED;
-                active_vpn_connection.vpn_state_changed.connect (() => {
-                    secure = active_vpn_connection.get_vpn_state () == NM.VpnConnectionState.ACTIVATED;
-                });
+            unowned string connection_type = ac.get_connection_type ();
+            if (connection_type == NM.SettingVpn.SETTING_NAME) {
+                /* We cannot rely on the sole state_changed signal, as it will
+                 * silently ignore sub-vpn specific states, like tun/tap
+                 * interface connection etc. That's why we keep a separate
+                 * implementation for the signal handlers. */
+                var _connection = (NM.VpnConnection) ac;
+                secure = secure || (_connection.get_vpn_state () == NM.VpnConnectionState.ACTIVATED);
+                _connection.vpn_state_changed.disconnect (update_vpn_connection);
```

Comment:
Are these disconnects helping? I feel like I am missing something here.

Reply:
Good question… I’m not sure how signaling really works internally. Are callbacks stored in a simple list or a set (i.e. if we connnect again and again the same callback, do we get a full list of duplicated handlers, or does the "connect" method already take care to discard existing handlers when registering? I cannot found anything about this in the documentation, that’s why I did the disconnection myself to ensure we keep only one callback at a time.

Actually, when I did my tests, I discover that without disconnecting previous handlers, I ended with a lot of parallel callback calls after some time (easily testable with a simple log output), that’s why I added this disconnect call before to avoid duplicate. But maybe I missed something.

If one has a good documentation entry about maintaining signal handlers, I’d appreciate!

Reply:
I think @Marukesu has made good suggestions for documentation before on the Slack. Maybe he has insight with this?